core/bin/dust_api.rs

```diff
@@ -2091,29 +2092,43 @@ async fn tables_upsert(
             &payload.parents,
             payload.remote_database_table_id,
             payload.remote_database_secret_id,
-        ),
-        // Only upsert the node if the title and mime_type are present.
-        match &maybe_ds_node {
-            Some(n) => state.store.upsert_data_source_node(&data_source_id, n),
-            None => Box::pin(futures::future::ok(())),
-        },
-    )
-    .await
+        )
+        .await
     {
-        Ok((table, _)) => (
-            StatusCode::OK,
-            Json(APIResponse {
-                error: None,
-                response: Some(json!({ "table": table })),
-            }),
-        ),
-        Err(e) => error_response(
-            StatusCode::INTERNAL_SERVER_ERROR,
-            "internal_server_error",
-            "Failed to upsert table",
-            Some(e),
-        ),
+        Ok(table) => table,
+        Err(e) => {
+            return error_response(
+                StatusCode::INTERNAL_SERVER_ERROR,
+                "internal_server_error",
+                "Failed to upsert table",
+                Some(e),
+            )
+        }
+    };
+
+    // Upsert the data source node if title and mime_type are present
+    if let Some(n) = &maybe_ds_node {
```

Comment:
Eventually (when title and mimeType are required) we might want to run this from the `upsert_table` function directly right?

Reply:
Maybe but now obvious. I think for now the paradigm is that store functions manage one specific table.

If we do it like this then we should probably remove entirely the store functions to manage data source nodes directly (and only expose functions to manage tables, docs and folders)

I don't really have an opinion on this yet

Reply:
We could add an upsert function on Table that calls the 2 separate store functions but using a shared transaction?

Reply:
Not a big deal anyway. It's just that we are open to having some tables without nodes if the node upsert fails here...

Reply:
Yeah we can def have that it's a good idea (even though technically we only have a single place where we upsert tables so we could pontially share the Tx in the handler).

Still I think having that on `Table` is a cleaner

Reply:
Only caveat is that in front we're moving towards the invariant that transactions should only exist in resources... store.rs is our cheap resource equivalent here so we're going against the direction we're taking in front....

food for thoughts